sdk/core/azure-core/tests/test_serialization.py

```diff
@@ -2,10 +2,356 @@
+            else:
+                return _expand_dict(vars(obj))
+
+    except TypeError:
+        return obj
+
+
+def _expand_dict(d):
+    return dict((key, _expand_value(value)) for key, value in d.items())
+
+
+class SerializerMixin(object):
+    """Mixin that provides methods for representing a model as a dictionary"""
+
+    def to_dict(self):
+        return _expand_value(vars(self))
 
-from azure.core.serialization import NULL
 
 def test_NULL_is_falsy():
     assert NULL is not False
     assert bool(NULL) is False
-    assert NULL is NULL
\ No newline at end of file
+    assert NULL is NULL
+
+@pytest.fixture
+def json_dumps_with_encoder():
+    def func(obj):
+        return json.dumps(obj, cls=ComplexEncoder)
+    return func
+
+def test_bytes(json_dumps_with_encoder):
+    test_bytes = b"mybytes"
+    result = json.loads(json_dumps_with_encoder(test_bytes))
+    try:
+        assert base64.b64decode(result) == test_bytes  # Python 3
+    except TypeError:
+        assert result == test_bytes  # Python 2.7
+
+def test_byte_array_ascii(json_dumps_with_encoder):
+    test_byte_array = bytearray("mybytes", "ascii")
+    result = json.loads(json_dumps_with_encoder(test_byte_array))
+    assert base64.b64decode(result) == test_byte_array
+
+def test_byte_array_utf8(json_dumps_with_encoder):
+    test_byte_array = bytearray("mybytes", "utf-8")
+    result = json.loads(json_dumps_with_encoder(test_byte_array))
+    assert base64.b64decode(result) == test_byte_array
+
+def test_byte_array_utf16(json_dumps_with_encoder):
+    test_byte_array = bytearray("mybytes", "utf-16")
+    result = json.loads(json_dumps_with_encoder(test_byte_array))
+    assert base64.b64decode(result) == test_byte_array
+
+def test_dictionary_basic(json_dumps_with_encoder):
+    test_obj = {
+        "string": "myid",
+        "number": 42,
+        "boolean": True,
+        "list_of_string": [1, 2, 3],
+        "dictionary_of_number": {"pi": 3.14},
+    }
+    complex_serialized = json_dumps_with_encoder(test_obj)
+    assert json.dumps(test_obj) == complex_serialized
+    assert json.loads(complex_serialized) == test_obj
+
+def test_model_basic(json_dumps_with_encoder):
+    class BasicModel(SerializerMixin):
+        def __init__(self):
+            self.string = "myid"
+            self.number = 42
+            self.boolean = True
+            self.list_of_ints = [1, 2, 3]
+            self.dictionary_of_number = {"pi": 3.14}
+
+    expected = BasicModel()
+    expected_dict = {
+        "string": "myid",
+        "number": 42,
+        "boolean": True,
+        "list_of_ints": [1, 2, 3],
+        "dictionary_of_number": {"pi": 3.14},
+    }
+    assert json.loads(json_dumps_with_encoder(expected.to_dict())) == expected_dict
+
+def test_dictionary_datetime(json_dumps_with_encoder):
+    test_obj = {
+        "timedelta": timedelta(1),
+        "date": date(2021, 5, 12),
+        "datetime": isodate.parse_datetime('2012-02-24T00:53:52.780Z'),
```

Comment:
When the conversation starts regarding deserialization - we will want to make sure we support weird .NET timestamps (with extra decimal places). When that conversation happens - what we do to serialize round-tripped values will be important as well.